src/core/libmaven/EIC.cpp

```diff
@@ -124,29 +128,120 @@ EIC *EIC::eicMerge(const vector<EIC *> &eics)
+
+    // create a sparse identity matrix of size `n`
+    auto ident = MatrixXf::Identity(n, n).sparseView();
+
+    // compute approximate second derivative of the identity matrix
+    auto D = _diff(_diff(ident));
+
+    // create a weights vector of length `n` initially containing only ones
+    // for coefficients
+    auto w = VectorXf(ArrayXf::Ones(n));
+
+    // instantiate a Cholesky decomposition linear solver for sparse matrices
+    SimplicialCholesky<SparseMatrix<float>> solver;
+
+    // instantiate a vector for intensity and one to store baseline
+    VectorXf intensityVec = Map<VectorXf>(intensity.data(), n);
+    VectorXf baselineVec;
+
+    // helper matrices for `select` operation used when creating binary vectors
+    auto ones = VectorXf(MatrixXf::Ones(n, 1));
+    auto zeros = VectorXf(MatrixXf::Zero(n, 1));
+
+    // TODO: ideally this should converge to a point where the baseline does
+    // not change anymnore, but since we do not have good float comparators
+    // yet we can use a decent number of iterations to get as close to the true
+    // baseline as possible
+    for (int i=0; i < numIterations; ++i) {
+        // create a square matrix 'W' with the weights as its diagonal
+        auto W = SparseMatrix<float>(w.asDiagonal());
+
+        // compute 'A' and 'b', and then solve for 'x' that satisfies the
+        // equation 'A·x = b', where x will be the iteratively estimated
+        // baseline
+        auto A = W + (lambda * (D.transpose() * D));
+        solver.compute(A);
+        auto b = VectorXf(w.array() * intensityVec.array());
+        baselineVec = solver.solve(b);
+
+        // calculate weights for the next iteration
+        auto gtBin = VectorXf(
+            ((intensityVec - baselineVec).array() > 0.0f).select(ones, zeros)
+        );
+        auto ltBin = VectorXf(
+            ((intensityVec - baselineVec).array() < 0.0f).select(ones, zeros)
+        );
+        w = (p * gtBin) + ((1.0f - p) * ltBin);
     }
 
+    // copy data from an Eigen vector to std::vector
+    vector<float> clippedBaseline(baselineVec.data(),
+                                  baselineVec.data()
+                                  + baselineVec.rows()
+                                  * baselineVec.cols());
+
+    // clip negative values from the baseline vector
+    std::transform(begin(clippedBaseline),
+                   end(clippedBaseline),
+                   begin(clippedBaseline),
+                   [](float f) { return f < 0.0f ? 0.0f : f; });
+
+    // since baseline right now is an array of float, the data from clipped
+    // vector is copied to this array
+    std::copy(begin(clippedBaseline), end(clippedBaseline), baseline);
+}
+
+Eigen::SparseMatrix<float> EIC::_diff(Eigen::SparseMatrix<float> mat)
+{
+    Eigen::SparseMatrix<float> E1 = mat.block(0, 0, mat.rows() - 1, mat.cols());
+    Eigen::SparseMatrix<float> E2 = mat.block(1, 0, mat.rows() - 1, mat.cols());
+    return E2 - E1;
```

Comment:
Since this is only ever used by the `_computeAsLSBaseline` method, why not just put this piece of code within the method as a lambda function. I don't see any reason why something so unrelated should be left hanging around as a member of our EIC class. What do you think?

Reply:
> why not just put this piece of code within the method as a lambda function.

Yes,we can do that or maybe declare it inline. Whatever seems appropriate to you